### ISSUE TITLE
Some annotations include stop codon in ORF

### ISSUE THREAD

**Issue by lvclark** [kvittingseerup/IsoformSwitchAnalyzeR#77]
Me again working with honeybee: https://www.ncbi.nlm.nih.gov/genome/?term=apis+mellifera

Unless I run `extractSequence` manually with `removeORFwithStop = TRUE`, all of my protein sequences get removed for supposedly having a stop codon before the end of the annotated CDS.  Actually, the annotated CDS includes the stop codon at the end.  It might be possible to trim this out with:

```
transcriptORFaaSeq <- Biostrings::trimLRpattern(Rpattern = "*", subject = transcriptORFaaSeq)
```

However, I'm not sure if the data in the `$orfAnalysis` slot also need to be adjusted for other downstream analysis.

Then there are a few others (e.g. NM_001011581.1) that are getting discarded because either the annotation or the transcript FASTA is just wrong, but that seems to be a problem with the annotation rather than your software.

**Comment by kvittingseerup** [comment]
Hi again - and thanks for reaching out again.

So all of your CDS regions include the stop codon? Btw where did you get the gene annotation from?

Cheers
Kristoffer

**Comment by lvclark** [comment]
Yes, all of the CDS included the stop codon.  The transcriptome and annotation came from NCBI at the link I provided.  Thanks!

**Comment by lvclark** [comment]
Looks like maybe CDS exclude the stop codon in GTF, but include the stop codon in GFF3?

https://www.biostars.org/p/206362/

**Comment by kvittingseerup** [comment]
Ahh - I did not know about this. Thanks for letting me know.

And yes it also requires the `$orfAnalysis` to be modified which is not straight forward to do since the presence of introns means we cannot just subtract 3.

Could I get you to try and use [gffread](http://ccb.jhu.edu/software/stringtie/gff.shtml#gffread_ex) to convert the GTF to a GTF and see if it then works?

gffread is used as follows:

```
gffread a_gff_file.gff3 -T -o a_gtf_file.gtf
```

Cheers
Kristoffer

**Comment by lvclark** [comment]
I just tried `gffread` and it retains the stop codon in the output GTF!

What I ended up doing yesterday was running `analyzeORF` to overwrite the annotated ORFs, not only to remove the stop codons, but also because there was a small minority of transcripts that seemed to be annotated incorrectly (if I translated the supposed ORF sequence it was junk).  In the vast majority of cases, `analyzeORF` found the annotated ORF.

**Comment by lvclark** [comment]
My unit has our own script that we use to convert NCBI GFF3 to GTF, so I updated it to fill in transcript_id and trim stop codons.  I managed to create a GTF that works with IsoformSwitchAnalyzeR!

**Comment by kvittingseerup** [comment]
Thanks for letting me know. I dug into the NCBI documentation and they adhear to [this GFF definition](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md). Where the CDS includes stop codon (see "Note 5").

Would it be possible for you to share this script so other can benefit as well?

Cheers
Kristoffer

**Comment by lvclark** [comment]
Sure!  Here is the script, edited a bit to remove some other processing we do.

``` r
infile <- "myfile.gff" # edit this
outfile <- "myfile.gtf" # edit this

gff0 <- rtracklayer::import(infile)

# Get transcript name by ID
txpt_rows <- which(!is.na(gff0$transcript_id))
transcripts <- gff0$transcript_id[txpt_rows]
names(transcripts) <- gff0$ID[txpt_rows]
  
# Change the parent column from list into vector
parentcol <- rep(NA_character_, length(gff0))
hasparent <- which(lengths(gff0$Parent) > 0)
if(all(lengths(gff0$Parent[hasparent]) == 1)){
  parentcol[hasparent] <- unlist(gff0$Parent[hasparent])
} else {
  # if there are multiple parents, pick the first that has a txpt
  parentcol[hasparent] <-
    sapply(gff0$Parent[hasparent],
           function(x){
             x <- x[x %in% names(transcripts)]
             if(length(x) == 0){
               return(NA_character_)
             } else {
               return(x[1])
             }
           } )
}

# Propagate transcript_id to CDS and exons if necessary
cds_rows <- which(gff0$type == "CDS" & is.na(gff0$transcript_id)) # only CDS that need it filled in
cds_rows <- cds_rows[which(parentcol[cds_rows] %in% names(transcripts))] # need parent with transcript_id
if(length(cds_rows) > 0){
  gff0$transcript_id[cds_rows] <- transcripts[parentcol[cds_rows]]
}

exon_rows <- which(gff0$type == "exon" & is.na(gff0$transcript_id)) # exons that need to be filled in
exon_rows <- exon_rows[which(parentcol[exon_rows] %in% names(transcripts))] # need parent with transcript_id
if(length(exon_rows) > 0){
  gff0$transcript_id[exon_rows] <- transcripts[parentcol[exon_rows]]
}

# Fill in for those without transcript; gene_id if present, else product
if(!is.null(gff0$gene_id)){
  tofill <- which(gff0$type %in% c("CDS", "exon") & is.na(gff0$transcript_id))
  gff0$transcript_id[tofill] <- gff0$gene_id[tofill]
}
if(!is.null(gff0$product)){
  tofill <- which(gff0$type %in% c("CDS", "exon") & is.na(gff0$transcript_id))
  gff0$transcript_id[tofill] <- gff0$product[tofill]
}

# remove stop codon from CDS to comply with GTF format

# for exon-exon junctions in the middle of a stop codon
removerows <- integer(0)

# fill in for top strand
CDS_top <- which(as.logical(gff0$type == "CDS" & BiocGenerics::strand(gff0) == "+"))
CDS_top_split <- split(CDS_top, parentcol[CDS_top])
modifyrows <- integer(length(CDS_top_split))
tosubtract <- integer(length(CDS_top_split))
for(i in seq_along(CDS_top_split)){
  cds <- CDS_top_split[[i]]
  last <- cds[which.max(BiocGenerics::start(gff0[cds]))]
  if(BiocGenerics::width(gff0[last]) > 3){
    # whole stop codon is in this exon (vast majority of cases)
    modifyrows[i] <- last
    tosubtract[i] <- 3L
  } else {
    # stop codon goes into the previous exon
    tosubtract[i] <- 3L - BiocGenerics::width(gff0[last]) # remaining nucleotides to remove
    removerows <- c(removerows, last)
    cds <- cds[cds != last]
    modifyrows[i] <- cds[which.max(start(gff0[cds]))]
  }
}
BiocGenerics::end(gff0[modifyrows]) <- BiocGenerics::end(gff0[modifyrows]) - tosubtract

# fill in for bottom strand
CDS_bot <- which(as.logical(gff0$type == "CDS" & BiocGenerics::strand(gff0) == "-"))
CDS_bot_split <- split(CDS_bot, parentcol[CDS_bot])
modifyrows <- integer(length(CDS_bot_split))
tosubtract <- integer(length(CDS_bot_split))
for(i in seq_along(CDS_bot_split)){
  cds <- CDS_bot_split[[i]]
  last <- cds[which.min(BiocGenerics::start(gff0[cds]))]
  if(BiocGenerics::width(gff0[last]) > 3){
    # whole stop codon is in this exon (vast majority of cases)
    modifyrows[i] <- last
    tosubtract[i] <- 3L
  } else {
    # stop codon goes into the previous exon
    tosubtract[i] <- 3L - BiocGenerics::width(gff0[last]) # remaining nucleotides to remove
    removerows <- c(removerows, last)
    cds <- cds[cds != last]
    modifyrows[i] <- cds[which.min(BiocGenerics::start(gff0[cds]))]
  }
}
BiocGenerics::start(gff0[modifyrows]) <- BiocGenerics::start(gff0[modifyrows]) + tosubtract

if(length(removerows) > 0) gff0 <- gff0[-removerows] # CDS that only contain stop codon

rtracklayer::export(gff0, outfile, format = "gtf")
```

**Comment by kvittingseerup** [comment]
Thanks for the R script!

I'll try and make some updates to IsoformSwitchAnalyzeR to solve this problem. I'll let you know when they are online :-)

Cheers
Kristoffer

**Comment by kvittingseerup** [comment]
This was fixed in 1.13.06
